### ISSUE TITLE
Document implementation of `IUmbracoIndexingHandler`

### ISSUE THREAD

**Issue by bjarnef** [umbraco/UmbracoDocs#5003]
## What type of issue is it?
Missing documentation

## Describe the issue
There are some great documentation about Examine and custom indexing: https://docs.umbraco.com/umbraco-cms/reference/searching/examine

However I couldn't find much information about re-indexing and deleting documents in a custom index.

Searching the source code I found `IUmbracoIndexingHandler` though.
https://github.com/umbraco/Umbraco-CMS/blob/33adbf41fa1f5c5d0759c70a7116114107addf56/src/Umbraco.Infrastructure/Examine/ExamineUmbracoIndexingHandler.cs#L18

however the documentation doesn't mention this.

In a project I implemented this though:

```
internal class CourseIndexingHandler : IUmbracoIndexingHandler
{
	// the default enlist priority is 100
	// enlist with a lower priority to ensure that anything "default" runs after us
	// but greater that SafeXmlReaderWriter priority which is 60
	private const int EnlistPriority = 80;
	private readonly IBackgroundTaskQueue _backgroundTaskQueue;
	private readonly IValueSetBuilder<CourseDataIndexingModel> _contentValueSetBuilder;
	private readonly Lazy<bool> _enabled;
	private readonly IExamineManager _examineManager;
	private readonly ILogger<CourseIndexingHandler> _logger;
	private readonly IMainDom _mainDom;
	private readonly IProfilingLogger _profilingLogger;
	private readonly IPublishedContentValueSetBuilder _publishedContentValueSetBuilder;
	private readonly ICoreScopeProvider _scopeProvider;
	private readonly ICourseDataService _courseDataService;

	public CourseIndexingHandler(
		IMainDom mainDom,
		ILogger<CourseIndexingHandler> logger,
		IProfilingLogger profilingLogger,
		ICoreScopeProvider scopeProvider,
		IExamineManager examineManager,
		IBackgroundTaskQueue backgroundTaskQueue,
		IValueSetBuilder<CourseDataIndexingModel> contentValueSetBuilder,
		IPublishedContentValueSetBuilder publishedContentValueSetBuilder,
		ICourseDataService courseDataService)
	{
		_mainDom = mainDom;
		_logger = logger;
		_profilingLogger = profilingLogger;
		_scopeProvider = scopeProvider;
		_examineManager = examineManager;
		_backgroundTaskQueue = backgroundTaskQueue;
		_contentValueSetBuilder = contentValueSetBuilder;
		_publishedContentValueSetBuilder = publishedContentValueSetBuilder;
		_courseDataService = courseDataService;
		_enabled = new Lazy<bool>(IsEnabled);
	}

	/// <inheritdoc />
	public bool Enabled => _enabled.Value;

	/// <inheritdoc />
	public void DeleteIndexForEntity(int entityId, bool keepIfUnpublished)
	{
		var actions = DeferedActions.Get(_scopeProvider);
		if (actions != null)
		{
			actions.Add(new DeferedDeleteIndex(this, entityId, keepIfUnpublished));
		}
		else
		{
			DeferedDeleteIndex.Execute(this, entityId, keepIfUnpublished);
		}
	}

	/// <inheritdoc />
	public void DeleteIndexForEntities(IReadOnlyCollection<int> entityIds, bool keepIfUnpublished)
	{
		var actions = DeferedActions.Get(_scopeProvider);
		if (actions != null)
		{
			actions.Add(new DeferedDeleteIndex(this, entityIds, keepIfUnpublished));
		}
		else
		{
			DeferedDeleteIndex.Execute(this, entityIds, keepIfUnpublished);
		}
	}

	/// <inheritdoc />
	public void ReIndexForContent(IContent sender, bool isPublished)
	{
		var actions = DeferedActions.Get(_scopeProvider);
		if (actions != null)
		{
			actions.Add(new DeferedReIndexForContent(_backgroundTaskQueue, this, sender, _courseDataService, isPublished));
		}
		else
		{
			DeferedReIndexForContent.Execute(_backgroundTaskQueue, this, sender, _courseDataService, isPublished);
		}
	}

	/// <inheritdoc />
	public void ReIndexForMedia(IMedia sender, bool isPublished) => throw new NotImplementedException();

	/// <inheritdoc />
	public void ReIndexForMember(IMember sender) => throw new NotImplementedException();

	/// <inheritdoc />
	public void DeleteDocumentsForContentTypes(IReadOnlyCollection<int> removedContentTypes) => throw new NotImplementedException();

	/// <summary>
	///     Used to lazily check if Examine Index handling is enabled
	/// </summary>
	/// <returns></returns>
	private bool IsEnabled()
	{
		//let's deal with shutting down Examine with MainDom
		var examineShutdownRegistered = _mainDom.Register(release: () =>
		{
			using (_profilingLogger.TraceDuration<CourseIndexingHandler>("Examine shutting down"))
			{
				_examineManager.Dispose();
			}
		});

		if (!examineShutdownRegistered)
		{
			_logger.LogInformation(
				"Examine shutdown not registered, this AppDomain is not the MainDom, Examine will be disabled");

			//if we could not register the shutdown examine ourselves, it means we are not maindom! in this case all of examine should be disabled!
			Suspendable.ExamineEvents.SuspendIndexers(_logger);
			return false; //exit, do not continue
		}

		_logger.LogDebug("Examine shutdown registered with MainDom");

		var registeredIndexers =
			_examineManager.Indexes.OfType<IUmbracoIndex>().Count(x => x.EnableDefaultEventHandler);

		_logger.LogInformation("Adding examine event handlers for {RegisteredIndexers} index providers.",
			registeredIndexers);

		// don't bind event handlers if we're not suppose to listen
		if (registeredIndexers == 0)
		{
			return false;
		}

		return true;
	}

	#region Deferred Actions

	private class DeferedActions
	{
		private readonly List<DeferedAction> _actions = new();

		public static DeferedActions? Get(ICoreScopeProvider scopeProvider)
		{
			IScopeContext? scopeContext = scopeProvider.Context;

			return scopeContext?.Enlist("examineEvents",
				() => new DeferedActions(), // creator
				(completed, actions) => // action
				{
					if (completed)
					{
						actions?.Execute();
					}
				}, EnlistPriority);
		}

		public void Add(DeferedAction action) => _actions.Add(action);

		private void Execute()
		{
			foreach (DeferedAction action in _actions)
			{
				action.Execute();
			}
		}
	}

	/// <summary>
	///     An action that will execute at the end of the Scope being completed
	/// </summary>
	private abstract class DeferedAction
	{
		public virtual void Execute()
		{
		}
	}

	/// <summary>
	///     Re-indexes an <see cref="IContent" /> item on a background thread
	/// </summary>
	private class DeferedReIndexForContent : DeferedAction
	{
		private readonly IBackgroundTaskQueue _backgroundTaskQueue;
		private readonly IContent _content;
		private readonly ICourseDataService _coursedataService;
		private readonly CourseIndexingHandler _courseIndexingHandler;
		private readonly bool _isPublished;

		public DeferedReIndexForContent(IBackgroundTaskQueue backgroundTaskQueue,
			CourseIndexingHandler courseIndexingHandler, IContent content, ICourseDataService coursedataService, bool isPublished)
		{
			_backgroundTaskQueue = backgroundTaskQueue;
			_courseIndexingHandler = courseIndexingHandler;
			_content = content;
			_coursedataService = coursedataService;
			_isPublished = isPublished;
		}

		public override void Execute() =>
			Execute(_backgroundTaskQueue, _courseIndexingHandler, _content, _coursedataService, _isPublished);

		public static void Execute(IBackgroundTaskQueue backgroundTaskQueue,
			CourseIndexingHandler examineUmbracoIndexingHandler, IContent content, ICourseDataService courseDataService, bool isPublished)
			=> backgroundTaskQueue.QueueBackgroundWorkItem(cancellationToken =>
			{
				using ICoreScope scope =
					examineUmbracoIndexingHandler._scopeProvider.CreateCoreScope(autoComplete: true);

				// TODO: Get indexing model
				var data = courseDataService.GetCourseDataIndexingModels(content).GetAwaiter().GetResult().ToArray();

				if (data == null || data.Length == 0)
					return Task.CompletedTask;

				// for content we have a different builder for published vs unpublished
				// we don't want to build more value sets than is needed so we'll lazily build 2 one for published one for non-published
				var builders = new Dictionary<bool, Lazy<List<ValueSet>>>
				{
					//[true] = new(() => examineUmbracoIndexingHandler._contentValueSetBuilder.GetValueSets(data).ToList()),
					[false] = new(() => examineUmbracoIndexingHandler._contentValueSetBuilder.GetValueSets(data).ToList())
				};

				// This is only for content - so only index items for IUmbracoContentIndex (to exlude members)
				foreach (IUmbracoIndex index in examineUmbracoIndexingHandler._examineManager.Indexes
							 //.OfType<IUmbracoContentIndex>()
							 .OfType<CourseInstanceIndex>()
							 //filter the indexers
							 .Where(x => isPublished || !x.PublishedValuesOnly)
							 .Where(x => x.EnableDefaultEventHandler))
				{
					if (cancellationToken.IsCancellationRequested)
					{
						return Task.CompletedTask;
					}

					List<ValueSet> valueSet = builders[index.PublishedValuesOnly].Value;
					index.IndexItems(valueSet);
				}

				return Task.CompletedTask;
			});
	}

	private class DeferedDeleteIndex : DeferedAction
	{
		private readonly CourseIndexingHandler _courseIndexingHandler;
		private readonly int _id;
		private readonly IReadOnlyCollection<int>? _ids;
		private readonly bool _keepIfUnpublished;

		public DeferedDeleteIndex(CourseIndexingHandler courseIndexingHandler, int id,
			bool keepIfUnpublished)
		{
			_courseIndexingHandler = courseIndexingHandler;
			_id = id;
			_keepIfUnpublished = keepIfUnpublished;
		}

		public DeferedDeleteIndex(CourseIndexingHandler courseIndexingHandler,
			IReadOnlyCollection<int> ids, bool keepIfUnpublished)
		{
			_courseIndexingHandler = courseIndexingHandler;
			_ids = ids;
			_keepIfUnpublished = keepIfUnpublished;
		}

		public override void Execute()
		{
			if (_ids is null)
			{
				Execute(_courseIndexingHandler, _id, _keepIfUnpublished);
			}
			else
			{
				Execute(_courseIndexingHandler, _ids, _keepIfUnpublished);
			}
		}

		public static void Execute(CourseIndexingHandler courseIndexingHandler, int id,
			bool keepIfUnpublished)
		{
			foreach (IUmbracoIndex index in courseIndexingHandler._examineManager.Indexes
						 .OfType<IUmbracoIndex>()
						 .Where(x => x.PublishedValuesOnly || !keepIfUnpublished)
						 .Where(x => x.EnableDefaultEventHandler))
			{
				index.DeleteFromIndex(id.ToString(CultureInfo.InvariantCulture));
			}
		}

		public static void Execute(CourseIndexingHandler courseIndexingHandler,
			IReadOnlyCollection<int> ids, bool keepIfUnpublished)
		{
			foreach (IUmbracoIndex index in courseIndexingHandler._examineManager.Indexes
						 .OfType<IUmbracoIndex>()
						 .Where(x => x.PublishedValuesOnly || !keepIfUnpublished)
						 .Where(x => x.EnableDefaultEventHandler))
			{
				index.DeleteFromIndex(ids.Select(x => x.ToString(CultureInfo.InvariantCulture)));
			}
		}
	}

	#endregion
}
```

So when a `Course` node is published (and cache is refreshed) it does re-index the value set based on the updated model.

Most of the stuff at `#region Deferred Actions` seems to be something that could be re-used from Umbraco core, but isn't public.

I also wonder it is isn't specific related to content, media and member, is there another interface with a `ReIndexForDocument()` or similar method. Maybe another interface `IUmbracoIndexingHandler` should inherit from like `IUmbracoIndex` inherits from `IIndex`?

and maybe the reindex methods can benefits from the base `ReIndexForDocument()` method?

```
void ReIndexForContent(IContent sender, bool isPublished);

void ReIndexForMember(IMember member);

void ReIndexForMedia(IMedia sender, bool isPublished);
```

Similar for deletion of a document:

```
void DeleteIndexForEntity(int entityId, bool keepIfUnpublished);
```

**Comment by sofietoft** [comment]
Thanks for raising this @bjarnef !

In general we try not to document too much that goes beyond the scope of Umbraco, and a lot of the Examine stuff is right on the border with this 😅 
To build custom indexes is generally not an Umbraco thing, but I do see the overlaps you're also highlighting here with the `IUmbracoIndexingHandler`.

I'll bring it to our developers to hear what they have to say.

**Comment by bjarnef** [comment]
I also wonder if `DeleteIndexForEntity` and `DeleteIndexForEntities` should pass in `string` instead of `int`. I know Umbraco use node id as id, but in custom indexes it is valid to use any other string value, like a guid or code.

![image](https://user-images.githubusercontent.com/2919859/223673087-395d6f2e-4ddc-4452-aba5-88bbc1a4b5d6.png)

Alternatively a new interface which `IUmbracoIndexingHandler` inherits from, which still can use `int entityId` but the new interface has a method like `DeleteDocument(string id)`.

**Comment by bjarnef** [comment]
@bergmania @bielu maybe something to consider in abstraction on Examine and if it need to introduce breaking changes anyway?
https://github.com/umbraco/Umbraco-CMS/pull/13851

E.g. the methods in `IUmbracoIndexingHandler` expect the documents in index to have `int` ids, because of Umbraco node/entity id. However in a custom index it can be any string value like a guid (eventually with hyphen removed).

I guess a custom index can possible be built based on data in [Konstrukt](https://marketplace.umbraco.com/package/konstrukt) or other data source where the key may not be an `int` id.

**Comment by bielu** [comment]
@bjarnef there is no way to change int id to string id without breaking change, it might be worth to include it in new abstraction for examine/ elastic/ algolia/ whatever :) as we decouple depedency on specific search engine :)

**Comment by bjarnef** [comment]
@bielu yes, that's why I mentioned it, when https://github.com/umbraco/Umbraco-CMS/pull/13851introduce breaking changes anyway, but not sure if in should happen in v13 or another future release.

It seems the document `DocId` is an `int` value https://lucene.apache.org/core/8_4_1/luke/org/apache/lucene/luke/models/documents/Documents.html but the id paramter in `ValueSet()` is a string.. which seems to set `__NodeId`.

It messed it a bit up if setting `__NodeId` in values dictionary as well.

**Comment by bergmania** [comment]
Hi @bjarnef.

I'm not sure I understand this correct, without knowing what a "Course" is. If this is some external data that is not related to `IContent`, I cant see why not just go plain examine, and avoid looking into what Umbraco is doing. I usually  look at https://github.com/umbraco/UmbracoExamine.PDF to see a simple setup using custom index.

Here Reindex and deletes are handled by the `PdfIndexPopulator`

**Comment by bjarnef** [comment]
Hi @bergmania 

`Course` is in this case a node, but besides the `Course` node, there also instances related to the course. The instance data doesn't exist as Umbraco data, but in a custom index. On full indexing these documents are built from a data source, but also combined with some data from the Course node, e.g. categories set on a Course node. But if these changes and Course node is published, we need to re-index these documents.

Originally I also tried a multi-searcher combining `ExternalIndex` and the custom index, but we need the search/filter to work on both "course" and "instance" data, even when some course don't have any instances.

The re-indexing using `IUmbracoIndexingHandler` does work though, but I may have a look at how it is handled in the `PdfIndexPopulator`.

**Comment by bergmania** [comment]
@bjarnef, the usecase sounds quite similar to the PDF use cases, where something is required when media updates.

**Comment by bjarnef** [comment]
@bergmania I will try with the approach used in `PdfIndexPopulator`:
https://github.com/umbraco/UmbracoExamine.PDF/blob/v11/dev/src/UmbracoExamine.PDF/PdfIndexPopulator.cs#L71-L81

Just to be sure, the `AddToIndex` method is both responsible to adding new documents to index and update existing documents with new value set?

**Comment by bergmania** [comment]
@bjarnef, yeah, it looks like that :)

**Comment by bielu** [comment]
@bjarnef yes it is, I was debugging this method when rewriting abstractions :)

**Comment by bjarnef** [comment]
@bergmania tested it a bit using the approach in Examine PDF.

Difference seems to be that the implementation of `IUmbracoIndexingHandler` in Umbraco starts a background task, so the publish action finish right away, while reindexing happens in background.

The approach in Examine PDF in `AddToIndex` method fetch data directly from `IMedia`, but if the indexing model are based on some other data (e.g. in out use-case the indexing model is based on `IContent` combined with some external data), it has to wait on this to finish.

At the moment I have only a method `GetCourseDataIndexingModels()` used for full indexing, which may take a bit time since it contains 100K items..

But using `GetCourseDataIndexingModel(content)` where content is `IContent` Course node, would of course be much faster.

**Comment by bjarnef** [comment]
@bergmania I tested to approach in Examine PDF, which work well after some optimization.

I noticed this format: 
```
MyIndexPopulator : IndexPopulator
```

But other documentation and some indexes in core use this format:

```
MyIndexPopulator : IndexPopulator<MyIndex>
```

Whats's the difference?

I also noticed we have this in composer - are both needed?

```
builder.Services.AddSingleton<IIndexPopulator, MyIndexPopulator >();
builder.Services.AddSingleton<MyIndexPopulator>();
```

**Comment by bielu** [comment]
@bjarnef about 1 question:
MyIndexPopulator : IndexPopulator
if you check IndexPopulator<TIndex>, you can see it overrides IsRegistered to check the type of index and ensure it populates only to specific type of index: https://github.com/umbraco/Umbraco-CMS/blob/df8f13d6b85f73c68473c5f3f1df2e54c6f33586/src/Umbraco.Infrastructure/Examine/IndexPopulator.cs#L13C26-L26
about second question, my guess is to make available access to strong typed MyIndexPopulator, but I think it is wrong way of regitering as in this case we should most likely reuse instance between both registrations :)
